File: 7/java/Vehicle.java
/*
    }
 * The Vehicle Class.
 *
 * @author Dominic M.
 * @since 2024-10-28
 * @version 1.0
 */

/**
 * This is the Vehicle Class.
 */
public class Vehicle {

    /**
     * The color of car.
     */
    private static String color;

    /**
     * The speed of car.
     */
    private static int speed;

    /**
     * The max speed.
     */
    private static int maxSpeed;

    /**
     * This is the constructor.
     *
     * @param colour of car
     * @param maximumSpeed of car
     */
    public Vehicle(final String colour, final int maximumSpeed) {
        color = colour;
        speed = 0;
        maxSpeed = maximumSpeed;
    }

    /**
     * Getter for car color.
     *
     * @return string of car color
     */
    public String getColor() {
        return color;
    }

    /**
     * Sets the car color.
     *
     * @param newColor of car
     */
    public void setColor(final String newColor) {
        if (newColor != null) {
            color = newColor;
        } else {
            color = "silver";
        }
    }

    /**
     * Getter for car speed.
     *
     * @return speed of car
     */
    public int getSpeed() {
        return speed;
    }

    /**
     * Getter for maxSpeed
     *
     * @return max speed
     */
    public int getMaxSpeed() {
        return maxSpeed;
    }

    /**
     * This method accelerates the car.
     *
     * @param accelPower how much force is being put into acceleration
     * @param accelTime how long it is accelerating for
     */
    public void accelerate(final int accelPower, final int accelTime) {
        speed = (accelPower * accelTime) + speed;
        if (speed > maxSpeed) {
            speed = maxSpeed;
        }
    }

    /**
     * This method brakes the car.
     *
     * @param brakePower how much force is being put into the brakes
     * @param brakeTime how long it is braking for
     */
    public void brake(final int brakePower, final int brakeTime) {
        speed = speed - (brakePower * brakeTime);
        if (speed < 0) {
          speed = 0;
        }
    }
}

File: 9/ts/Vehicle.ts
/**
 * The Vehicle Class
 *
 * author Dominic M.
 * version 1.0
 * since 2024-10-28
 */

export class Vehicle {
  /**
   * The color of car
   */
  protected color: string

  /**
   * The speed of car
   */
  protected speed: number

  /**
   * The max speed
   */
  protected readonly maxSpeed: number

  /**
   * The amount of wheels in the vehicle
   */
  protected wheels: number

  /**
   * This is the constructor
   * 
   * @param color 
   * @param maxSpeed 
   * @param wheels 
   */
  constructor (color: string, maxSpeed: number, wheels: number) {
    this.color = color
    this.speed = 0
    this.maxSpeed = maxSpeed
    this.wheels = wheels
  }

  /**
   * Getter for car color
   *
   * @returns {string} of car color
   */
  public get theColor (): string {
    return this.color
  }

  /**
   * Sets the car color
   *
   * @param {string} newColor of car
   */
  public set theColor (newColor: string) {
    if (newColor != null) {
      this.color = newColor
    } else {
      this.color = 'silver'
    }
  }

  /**
   * Getter for car speed
   *
   * @returns {number} of car speed
   */
  public get theSpeed (): number {
    return this.speed
  }

  /**
   * Getter for max speed
   *
   * @returns {number} of max car speed
   */
  public get theMaxSpeed (): number {
    return this.maxSpeed
  }

  /**
   * Getter for wheels
   *
   * @return {number} of wheels
   */
  public get theWheels (): number {
    return this.wheels
  }

  /**
   * This method accelerates the car
   *
   * @param {number} accelPower
   * @param {number} accelTime
   */
  public accelerate (accelPower: number, accelTime: number): void {
    this.speed = (accelPower * accelTime) + this.speed
    if (this.speed > this.maxSpeed) {
      this.speed = this.maxSpeed
    }
  }

  /**
   * This method decelerates the car
   *
   * @param {number} breakPower
   * @param {number} breakTime
   */
  public break (breakPower: number, breakTime: number): void {
    this.speed = this.speed - (breakPower * breakTime)
    if (this.speed < 0) {
      this.speed = 0
    }
  }
}
